### ISSUE TITLE
postUpdate is not called when changing related collection from non-empty to empty

### ISSUE THREAD

**Issue by sidPL** [doctrine/orm#8075]
I was redirected here from symfony bugtraq as it concerns doctrine collections.

**Description**  

I have User entity with ManyToMany UserRoles collection. Also have EventSubscriber class which has postUpdate event subscribed. Everything is working fine until change on User is removing all objects from collection `$user->userRoles` - `postUpdate()` is not called and collection's `getDeleteDiff()` is empty. 

**How to reproduce**  
```
class User {
    /**
     * @ORM\ManyToMany(targetEntity="App\Entity\UserRole")
     */
    private $userRoles;
    (...other fields...)
}
```
- When changing non-relational fields - event is called. 
- When adding to `$user->userRoles` some roles - event is called (changes are visible using collection's `getInsertDiff()`)
- When removing from `$user->userRoles` - event is called (changes are visible using collection's `getDeleteDiff()`) but ONLY when after-update collection is not empty 
- When removing ALL roles from collection `$user->userRoles` - postUpdate event is NOT called
- When removing ALL roles from collection `$user->userRoles` AND changing some other field - event is called, changes of other field - visible, but there are no visible changes on `$user->userRoles` in `$user->userRoles->getDeleteDiff()`

**Doctrine using**
```
doctrine/annotations                v1.8.0
doctrine/cache                      1.10.0
doctrine/collections                1.6.4
doctrine/common                     2.12.0
doctrine/dbal                       v2.10.1
doctrine/doctrine-bundle            2.0.7
doctrine/doctrine-migrations-bundle 2.1.2
doctrine/event-manager              1.1.0
doctrine/inflector                  1.3.1
doctrine/instantiator               1.3.0
doctrine/lexer                      1.2.0
doctrine/migrations                 2.2.1
doctrine/orm                        v2.7.1
doctrine/persistence                1.3.6
doctrine/reflection                 v1.1.0
```

**Comment by hockdudu** [comment]
We had the same problem that the event isn't called.

In our case, I can confirm it's related to https://stackoverflow.com/questions/15311083/whats-the-proper-use-of-unitofwork-getscheduledcollectiondeletions-in-doctr.

`MergeDoctrineCollectionListener` (https://github.com/symfony/symfony/blob/6.1/src/Symfony/Bridge/Doctrine/Form/EventListener/MergeDoctrineCollectionListener.php) is the root of the problem, as when it's disabled (by disabled I mean commenting out the code inside the `vendor/` folder, as I couldn't find a way of disabling it properly), the listener works.

So, this seems to be a bug in Symfony, in the Package `symfony/doctrine-bridge`, and not in Doctrine.

**Comment by hockdudu** [comment]
Apparently, this is a long-standing bug. From what I could find, it was being tracked in https://github.com/symfony/symfony/issues/17154, but it was closed in favor of https://github.com/doctrine/orm/pull/7120, which was also closed after inactivity.

Numerous other issues reference this problem, but all were closed for one reason or another
- https://github.com/symfony/symfony/issues/13713
- https://github.com/symfony/symfony/issues/34120
- https://github.com/symfony/symfony/issues/11262

**Comment by greg0ire** [comment]
> So, this seems to be a bug in Symfony, in the Package symfony/doctrine-bridge, and not in Doctrine.

Ok, thanks for investigating! I guess you should open a new Symfony issue?

**Comment by hockdudu** [comment]
@greg0ire Yes, I'll open a new Symfony issue.

**Comment by beberlei** [comment]
I am not sure only symfony is at fault. However some collection changes are detected differently than changesets and thst is what counts for postUpdste. Its a tricky implementation detail. I thought it was even documented but its only hinted at

**Comment by GSpecDrum** [comment]
@hockdudu Have you opened a new issue in Symfony repo?

**Comment by GSpecDrum** [comment]
I found that if set `'by_reference' => false,` option to EntityType form, then UnitOfWork detect changes of collection. 

Because of
https://github.com/symfony/symfony/blob/6.1/src/Symfony/Component/Form/Form.php#L302
and 
https://github.com/doctrine/orm/blob/2.12.x/lib/Doctrine/ORM/PersistentCollection.php#L618
->
`$this->isDirty = true;`

(Method in entity for removing item from collection must be exist)

**Comment by hockdudu** [comment]
@GSpecDrum No, I didn't open any issue yet, as I wanted to create a new repo where the problem may be reproduced, but I hadn't come to it yet.

I had some spare time now, so I experimented on a new Symfony application. You are totally correct with both your statements:

1. `by_reference` must be set to `false`. If this isn't set, then the event handler won't be called when the collection is emptied.
2. The entity must have the respective `addXXX` and `removeXXX` methods. If they don't exist (e.g. you have a `setXXX` instead), then the event is called, but `->getDeleteDiff()` will be empty when the collection is emptied.

With both requirements fulfilled, I wasn't able to reproduce this issue anymore.

When I originally faced this issue, I tried setting `by_reference` to `false`, but I didn't realize that the method `removeXXX` is _required_ for this to work (we only had a `setXXX` method). I went back to the project this issue occurred, added the `removeXXX` method, and now it works!

With this, I believe there is no need to opening an issue anymore.
